src/forward_proxy.rs

```diff
@@ -0,0 +1,96 @@
+use crate::errors::*;
+use nix::sys::socket;
+use std::net::SocketAddr;
+use std::os::unix::io::AsRawFd;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpListener, TcpStream};
+
+use crate::reverse_proxy;
+
+pub const PROXY_REDIR_PORT: u16 = 8080;
+
+pub fn run(local_addr: SocketAddr, compress: bool, encrypt: bool) -> Result<()> {
+    let rt = tokio::runtime::Runtime::new().chain_err(|| "failed to create tokio runtime")?;
+
+    rt.block_on(async {
+        println!("opening listener socket on {}", local_addr);
+        let listen_socket = TcpListener::bind(local_addr)
+            .await
+            .chain_err(|| format!("error opening listener socket on {}", local_addr))?;
+
+        socket::setsockopt(
+            listen_socket.as_raw_fd(),
+            socket::sockopt::IpTransparent,
+            &true,
+        )?;
+
+        loop {
+            let (from_conn, from_addr) = listen_socket
+                .accept()
+                .await
+                .chain_err(|| format!("error accepting connection"))?;
+            println!("connection received from {}", from_addr);
+
+            match socket::getsockname(from_conn.as_raw_fd()) {
+                Ok(socket::SockAddr::Inet(inet_addr)) => {
+                    println!("connection destined to {}", inet_addr);
+
+                    let to_addr =
+                        SocketAddr::new(inet_addr.ip().to_std(), reverse_proxy::HTTPS_PORT);
```

Comment:
Can we also pull the original destination port from the connection too?

Reply:
The original destination port is always going to be http 80 (9980 in our case) because the server configuration in script/tproxy_iptables.sh only redirects connections to that port over to our 8080 proxy listening port. Theoretically, we could, but that might involve redirecting connections to all of forward-proxy's ports over to 8080 instead of just http connections.